spec/app/domain/authentication/authn-oidc/discover_oidc_provider_spec.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+RSpec.describe Authentication::AuthnOidc::AuthenticateIdToken::DiscoverOIDCProvider do
+
+  let (:test_provider_uri) { "test-provider-uri" }
+  let (:test_error) { "test-error" }
+  let (:mock_provider) { "test-provider" }
+
+  def mock_discovery_provider(error:)
+    double('discovery_provider').tap do |discovery_provider|
+      if error
+        allow(discovery_provider).to receive(:discover!)
+                                       .and_raise(error)
+      else
+        allow(discovery_provider).to receive(:discover!)
+                                       .and_return(mock_provider)
+      end
+    end
+  end
+
+  context "A discoverable Oidc provider" do
+    subject do
+      Authentication::AuthnOidc::AuthenticateIdToken::DiscoverOIDCProvider.new(
+        open_id_discovery_service: mock_discovery_provider(error: nil)
+      ).(
+        provider_uri: test_provider_uri
+      )
+    end
+
+    it "does not raise an error" do
+      expect { subject }.to_not raise_error
+    end
+
+    it "returns the discovered provider" do
+      expect(subject).to eq(mock_provider)
+    end
+  end
+
+  context "A non-discoverable Oidc provider" do
+    context "fails on timeout error" do
+      subject do
+        Authentication::AuthnOidc::AuthenticateIdToken::DiscoverOIDCProvider.new(
+          open_id_discovery_service: mock_discovery_provider(error: HTTPClient::ConnectTimeoutError)
```

Comment:
A followup comment to the above one...

Notice here that something funny is going on... our tests know about `HTTPClient::ConnectTimeoutError`, but the section of code in `DiscoverOidcProvider` where `HTTPClient::ConnectTimeoutError` is used is *private*.  

Yet as we know unit tests are only supposed to test public behavior.

If `Authentication::AuthnOidc::AuthenticateIdToken::DiscoverOIDCProvider.discover!` were instead wrapped in an adapter class we owned, then the error translation to our custom errors would be done there, in that adapter class.  And here, where we're testing our application code, you'd be injecting a mock of the adapter class which directly raised our custom `Errors::Authentication::AuthnOidc::ProviderDiscoveryTimeout`, rather than raising the `HttpClient` error which *we happen to know* (even though we shouldn't, because it's private info) gets translated into that.